python/ray/tune/ray_trial_executor.py

```diff
@@ -436,11 +436,26 @@ def save(self, trial, storage=Checkpoint.DISK):
         """Saves the trial's state to a checkpoint."""
         trial._checkpoint.storage = storage
         trial._checkpoint.last_result = trial.last_result
+
+        subdir = ""
+        if trial.keep_best_checkpoint:
+            mean_rew_since_checkpoint = sum(trial.results_since_checkpoint) / len(trial.results_since_checkpoint)
+            if mean_rew_since_checkpoint > trial.past_avg:
+                subdir = "best/"
+                trial.past_avg = mean_rew_since_checkpoint
+            trial.results_since_checkpoint = []
+
         if storage == Checkpoint.MEMORY:
             trial._checkpoint.value = trial.runner.save_to_object.remote()
         else:
             with warn_if_slow("save_to_disk"):
-                trial._checkpoint.value = ray.get(trial.runner.save.remote())
+                trial._checkpoint.value, folder_path = ray.get(trial.runner.save_checkpoint_relative.remote(subdir))
+
+            if trial.prefix[subdir]["limit"]:
+                if len(trial.prefix[subdir]["history"]) == trial.prefix[subdir]["limit"]:
+                    ray.get(trial.runner.delete_checkpoint_relative.remote(trial.prefix[subdir]["history"][-1]))
+                    trial.prefix[subdir]["history"].pop()
+                trial.prefix[subdir]["history"].insert(0, folder_path)
```

Comment:
use dequeue instead, it is not as important but this collection is not optimized for inserting at the begining.

Reply:
Deque is not json serializable which is needed for this class